### ISSUE TITLE
Read failed; Trying to read beyond buffer size

### ISSUE THREAD

**Issue by parasAtmedia** [TileDB-Inc/TileDB-Py#36]
1. I had created KV array of tiledb, and stored around 10000 key-value pairs (tiledb directory sizing upto 18MB)
2. I had consolidated KV array after writes
3. I started to read the KV array based on those 10000 keys (just to get a benchmark on read performance), but it throwed an error
```
File "key_value.py", line 55, in store_vectors
    r = self.key_value_instance[v["key"]]
  File "tiledb/libtiledb.pyx", line 1567, in tiledb.libtiledb.KV.__getitem__
  File "tiledb/libtiledb.pyx", line 120, in tiledb.libtiledb.check_error
  File "tiledb/libtiledb.pyx", line 116, in tiledb.libtiledb._raise_ctx_err
  File "tiledb/libtiledb.pyx", line 101, in tiledb.libtiledb._raise_tiledb_error
tiledb.libtiledb.TileDBError: [TileDB::Buffer] Error: Read failed; Trying to read beyond buffer size
```
I don't get this error if I skip step 2 (consolidation of the updates), but my read performance drops to around 1.5 seconds for each key read

**Comment by stavrospapadopoulos** [comment]
Hi there,

We will certainly look into it. A couple of questions:

1) What version of `libtiledb` does the Python wrapper use?
2) Can you please share some code on how you create and populate the kv store? On the C side there is a function `tiledb_kv_set_max_buffered_items` that essentially controls the number of fragments to be created. If you set it >10,000 in you example, you will end up with a single fragment, thus no consolidation will be needed (if you need a quick solution until we see what the problem with consolidation is). @jakebolewski is that exposed in the Python API?

**Comment by jakebolewski** [comment]
Unfortunately no, the KV integration is not very optimized / inefficient currently.  We can push an update that fixes these issues.

**Comment by parasAtmedia** [comment]
Sorry, was not able to respond soon.
1. tiledb version is 0.1.1
2. Sample Code:
```
self.tile_db_context = tiledb.Ctx()
self.a1 = tiledb.Attr(self.tile_db_context, "attribute_value", compressor=("gzip", -1), dtype=bytes)
self.key_value_instance = tiledb.KV(self.tile_db_context, "array_name", attrs=(self.a1,))
self.key_value_instance.dump()

index = 0
data_dic = {}
# self.data_set is a list of dictionaries {"key": ..., "value":...}
for v in self.data_set:
    data_dic[v["key"]] = v["value"]
    index += 1
    if index % 100 == 0:
        self.key_value_instance.update(data_dic)
        data_dic = {}

self.key_value_instance.consolidate()
```

I am performing kv update after every 100 instances. Update after each key-value pair was slowing it down.

**Comment by jakebolewski** [comment]
d363292b8ab479f07ab0 should help with the write performance and the need to consolidate, unfortunately read performance will still be an issue until the next version of TileDB, the sparse read algorithm is being overhauled.

This will appear in an upcoming patch release of TileDB-Py

**Comment by parasAtmedia** [comment]
Thanks for responding. Looking forward to 0.1.2!

**Comment by jakebolewski** [comment]
All these fixes will be integrated into TileDB 1.3 / TileDB-Py 0.2, which will be released next week.
